mungegithub/mungers/approval-handler.go

```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+		return err
+	}
+	latestNotification := mungeComment.FilterComments(comments, notificationMatcher).GetLast()
+	latestApprove := mungeComment.FilterComments(comments, mungeComment.CommandName("approve")).GetLast()
+	if latestNotification == nil {
+		return h.createMessage(obj, needsApproval)
+	}
+	if latestApprove == nil {
+		// there was already a bot notification and nothing has changed since
+		return nil
+	}
+	if latestApprove.CreatedAt.After(*latestNotification.CreatedAt) {
+		// there has been approval since last notification
+		obj.DeleteComment(latestNotification)
+		return h.createMessage(obj, needsApproval)
+	}
+	lastModified := obj.LastModifiedTime()
+	if latestNotification.CreatedAt.Before(*lastModified) {
+		obj.DeleteComment(latestNotification)
+		return h.createMessage(obj, needsApproval)
+	}
+	return nil
+}
 
-		if len(fields) == 1 && strings.ToLower(fields[0]) == "/approve" {
-			approverSet.Insert(*c.User.Login)
-			continue
+// findApproverSet Takes all the Owners Files that Are Needed for the PR and chooses a good
+// subset of Approvers that are guaranteed to be from all of them (exact cover)
+// This is a greedy approximation and not guaranteed to find the minimum number of OWNERS
+func (h ApprovalHandler) findApproverSet(ownersPath sets.String) sets.String {
+
+	// approverCount contains a map: person -> set of relevant OWNERS file they are in
+	approverCount := make(map[string]sets.String)
+	for ownersFile := range ownersPath {
+		for approver := range h.features.Repos.LeafApprovers(ownersFile) {
+			if _, ok := approverCount[approver]; ok {
+				approverCount[approver].Insert(ownersFile)
+			} else {
+				approverCount[approver] = sets.NewString(ownersFile)
+			}
 		}
+	}
 
-		if len(fields) == 2 && strings.ToLower(fields[0]) == "/approve" && strings.ToLower(fields[1]) == "cancel" {
-			approverSet.Delete(*c.User.Login)
+	var copyOfFiles sets.String
+	for fn := range ownersPath {
+		copyOfFiles.Insert(fn)
+	}
+
+	approverGroup := sets.NewString()
+	for copyOfFiles.Len() > 0 {
+		maxCovered := 0
+		var bestPerson string
+		for k, v := range approverCount {
+			if len(v) >= maxCovered && v.Intersection(copyOfFiles).Len() != 0 {
+				maxCovered = len(v)
+				bestPerson = k
+			}
 		}
+
+		approverGroup.Insert(bestPerson)
+		copyOfFiles.Delete(approverCount[bestPerson].List()...)
 	}
+	return approverGroup
+}
 
+func (h *ApprovalHandler) createMessage(obj *github.MungeObject, filesNeedApproval sets.String) error {
+	files, err := obj.ListFiles()
+	if err != nil {
+		glog.Error("Could not list the files for PR%v", obj.Issue.Number)
+		return err
+	}
+	// contains the a set of directories where relevant ownersFiles are
+	ownersFiles := sets.String{}
 	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+		// find owners files for files that have not yet been approved
+		if !filesNeedApproval.Has(*file.Filename) {
```

Comment:
So if a file is NOT in `filesNeedApproval`, we looks for it `OWNERS` path? that seems wrong.